core/src/main/java/com/linecorp/armeria/client/ClientConnectionTimingsBuilder.java

```diff
@@ -0,0 +1,122 @@
+        checkState(!dnsResolutionEndSet, "dnsResolutionEnd() is already called.");
+        dnsResolutionEndNanos = System.nanoTime();
+        dnsResolutionEndSet = true;
+        return this;
+    }
+
+    /**
+     * Sets the time when connecting to a remote peer started.
+     */
+    public ClientConnectionTimingsBuilder socketConnectStart() {
+        socketConnectStartTimeMicros = SystemInfo.currentTimeMicros();
+        socketConnectStartNanos = System.nanoTime();
+        return this;
+    }
+
+    /**
+     * Sets the time when connecting to a remote peer ends.
+     *
+     * @throws IllegalStateException if {@link #socketConnectStart()} is not invoked before calling this.
+     */
+    public ClientConnectionTimingsBuilder socketConnectEnd() {
+        checkState(socketConnectStartTimeMicros >= 0, "socketConnectStart() is not called yet.");
+        checkState(!socketConnectEndSet, "socketConnectEnd() is already called.");
+        socketConnectEndNanos = System.nanoTime();
+        socketConnectEndSet = true;
+        return this;
+    }
+
+    /**
+     * Sets the time when waiting the completion of an existing connection attempt in order to use one
+     * connection for HTTP/2.
+     */
+    public ClientConnectionTimingsBuilder pendingAcquisitionStart() {
+        pendingAcquisitionStartTimeMicros = SystemInfo.currentTimeMicros();
+        pendingAcquisitionStartNanos = System.nanoTime();
+        return this;
+    }
+
+    /**
+     * Sets the time when waiting an existing connection attempt ends in order to use one connection
+     * for HTTP/2.
+     *
+     * @throws IllegalStateException if {@link #pendingAcquisitionStart()} is not invoked before calling this.
+     */
+    public ClientConnectionTimingsBuilder pendingAcquisitionEnd() {
+        checkState(pendingAcquisitionStartTimeMicros >= 0, "pendingAcquisitionStart() is not called yet.");
+        checkState(!pendingAcquisitionEndSet, "pendingAcquisitionEnd() is already called.");
+        pendingAcquisitionEndNanos = System.nanoTime();
+        pendingAcquisitionEndSet = true;
+        return this;
+    }
+
+    /**
+     * Returns a newly-created {@link ClientConnectionTimings} instance.
+     */
+    public ClientConnectionTimings build() {
+        return new ClientConnectionTimings(
+                connectionAcquisitionStartTimeMicros,
+                System.nanoTime() - connectionAcquisitionStartNanos,
+                dnsResolutionEndSet ? connectionAcquisitionStartTimeMicros : -1,
+                dnsResolutionEndSet ? dnsResolutionEndNanos - connectionAcquisitionStartNanos : -1,
+                socketConnectEndSet ? socketConnectStartTimeMicros : -1,
+                socketConnectEndSet ? socketConnectEndNanos - socketConnectStartNanos : -1,
+                pendingAcquisitionEndSet ? pendingAcquisitionStartTimeMicros : -1,
+                pendingAcquisitionEndSet ? pendingAcquisitionEndNanos - pendingAcquisitionStartNanos : -1);
```

Comment:
Should we log a warning or raise an exception if we recorded start time but not the end time?

Reply:
I think a warning is enough. Added, thanks!